### PR TITLE
AfterEffects: Implemented New Publisher

### PR DESCRIPTION
**Pull request by kalisp**
## Brief description
Implementation of New Publisher in AfterEffects

## Description
Implementation should allow use of old and New publisher at same time

PR contains https://github.com/pypeclub/OpenPype/pull/2841 , after that is merged, list of modified files should be much smaller.

Known issue:
- old publisher shows all 3 creators, will be fixed with https://github.com/pypeclub/OpenPype/pull/2854

## Testing notes:
- Enable New Publisher in Local Settings
- Test full create > publish process on empty workfile
- Test publishing in old workfile with created instances in NP and via old publish

### CODE REVIEW

**Comment by mkolar** [comment]
Task linked: [OP-2765 AE to new publisher](https://app.clickup.com/t/6658547/OP-2765)

File: openpype/hosts/aftereffects/api/pipeline.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from openpype.api import Logger
 from openpype.pipeline import LegacyCreator
 import openpype.hosts.aftereffects
+from openpype.pipeline import BaseCreator
```
Comment (iLLiCiTiT) [inline openpype/hosts/aftereffects/api/pipeline.py:14]:
```suggestion
from openpype.pipeline import LegacyCreator, BaseCreator
import openpype.hosts.aftereffects
```

**Comment by mkolar** [comment]
Before this is merged we need to Create user documentation for the new publisher.

File: openpype/hosts/aftereffects/plugins/create/create_render.py [hunk 2]
```diff
@@ -1,61 +1,118 @@
...
+        for inst in self.create_context.instances:
+            if subset_name == inst.subset_name:
+                raise CreatorError("{} already exists".format(
+                    inst.subset_name))
+
+        data["members"] = [items[0].id]
+        new_instance = CreatedInstance(self.family, subset_name, data, self)
+        new_instance.creator_attributes["farm"] = pre_create_data["farm"]
+
+        api.get_stub().imprint(new_instance.get("instance_id"),
```
Comment (iLLiCiTiT) [inline openpype/hosts/aftereffects/plugins/create/create_render.py:75]: You can use `id` attribute on instance object.
```suggestion
        api.get_stub().imprint(new_instance.id,
```

File: openpype/hosts/aftereffects/plugins/create/create_render.py [hunk 3]
```diff
@@ -1,61 +1,118 @@
...
-            instance['subset'].lower()
-            for instance in list_instances()
+        for inst in self.create_context.instances:
+            if subset_name == inst.subset_name:
+                raise CreatorError("{} already exists".format(
+                    inst.subset_name))
+
+        data["members"] = [items[0].id]
+        new_instance = CreatedInstance(self.family, subset_name, data, self)
+        new_instance.creator_attributes["farm"] = pre_create_data["farm"]
```
Comment (iLLiCiTiT) [inline openpype/hosts/aftereffects/plugins/create/create_render.py:73]: Pre create data may not be passed (e.g. when called from headless).
```suggestion
        if "farm" in pre_create_data:
            use_farm = pre_create_data["farm"]
            new_instance.creator_attributes["farm"] = use_farm
```

File: openpype/lib/delivery.py [hunk 4]
```diff
@@ -71,14 +83,15 @@ def path_from_representation(representation, anatomy):
 
 def copy_file(src_path, dst_path):
     """Hardlink file if possible(to save space), copy if not"""
-    from openpype.lib import create_hard_link  # safer importing
+    from avalon.vendor import filelink  # safer importing
```
Comment (iLLiCiTiT) [inline openpype/lib/delivery.py:86]: This is not right resolved.

File: openpype/plugins/publish/integrate_new.py [hunk 5]
```diff
@@ -220,7 +224,10 @@ def register(self, instance):
 
         # Ensure at least one file is set up for transfer in staging dir.
         repres = instance.data.get("representations")
-        assert repres, "Instance has no files to transfer"
+        repres = instance.data.get("representations")
```
Comment (iLLiCiTiT) [inline openpype/plugins/publish/integrate_new.py:227]: duplicated line
```suggestion
```

File: openpype/hosts/aftereffects/plugins/create/create_render.py [hunk 6]
```diff
@@ -1,61 +1,120 @@
...
+    description = "Render creator"
 
-        Creates subsets in format 'familyTaskSubsetname',
-        eg 'renderCompositingMain'.
+    create_allow_context_change = False
 
-        Create only single instance from composition at a time.
-    """
+    def get_icon(self):
+        return resources.get_openpype_splash_filepath()
```
Comment (iLLiCiTiT) [inline openpype/hosts/aftereffects/plugins/create/create_render.py:22]: Probably should return different icon than OpenPype logo.

Comment (iLLiCiTiT) [inline openpype/hosts/aftereffects/plugins/create/create_render.py:22]: Probably should return different icon than OpenPype logo.

File: openpype/lib/log.py [hunk 7]
```diff
@@ -493,3 +494,14 @@ def timed(*args, **kw):
...
+
+def debug_log_instance(logger, msg, instance):
+    """Helper function to write instance.data as json"""
+    def _default_json(value):
+        return str(value)
+
+    logger.debug(msg)
+    logger.debug(
+        json.dumps(instance.data, indent=4, default=_default_json)
+    )
```
Comment (iLLiCiTiT) [inline openpype/lib/log.py:507]: This is lib (and very specific) function in loggers logic. And I think it should not log but just convert the passed data into json format. Move it to `openpype/lib/config.py` (there are uncategorized lib functions now) and remove logging logic.

Example:
```
def _default_json_convert(value):
    return str(value)


def debug_data_dump(data, indent=4, convert_func=None):
    if convert_func is None:
        convert_func = _default_json_convert
    return json.dumps(data, indent=indent, default=convert_func)
```

File: openpype/hosts/aftereffects/api/pipeline.py [hunk 8]
```diff
@@ -15,6 +15,7 @@
     deregister_loader_plugin_path,
 )
 import openpype.hosts.aftereffects
+from openpype.pipeline import BaseCreator
```
Comment (iLLiCiTiT) [inline openpype/hosts/aftereffects/api/pipeline.py:18]: There already is import from `openpype.pipeline` at the top.

**Review by 64qam — APPROVED** [review]

**Pull request merged**